### PR TITLE
GH-1463: Add cilium identity list command

### PR DESCRIPTION
**Pull request by ashwinp**
- cilium identity get command should list individual identities.
- To be consistent with the other cilium API definitions, the
cilium identity list is the right command to display a list of
all/filtered identities.
- This patch adds the implementation of cilium identity list.
- Added new unit test for cilium identity list
- Added new unit test for cilium identity list --reserved
- Updated .gitignore

Signed-off-by: Ashwin Paranjpe <ashwin@covalent.io>

### CODE REVIEW

File: cilium/cmd/identity_get.go [hunk 1]
```diff
@@ -17,32 +17,24 @@ package cmd
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/cilium/cilium/pkg/policy"
-
-	identityApi "github.com/cilium/cilium/api/v1/client/policy"
 
 	"github.com/spf13/cobra"
```
Comment (aanm) [inline cilium/cmd/identity_get.go]: Imports order should be `golib`, `cilium`, `3rd parties`

File: cilium/cmd/identity_list.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/spf13/cobra"
```
Comment (aanm) [inline cilium/cmd/identity_list.go]: Imports order should be `golib`, `cilium`, `3rd parties`

File: .gitignore [hunk 3]
```diff
@@ -6,6 +6,7 @@
 # Folders
 _obj
 _test
+tests/cilium-files
```
Comment (aanm) [inline .gitignore:9]: @ianvernon I've spoked with @ashwinp it seems there are some tests files laying around. Can we remove them?

File: daemon/daemon_test.go [hunk 4]
```diff
@@ -22,8 +22,9 @@ import (
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/policy"
 
-	. "gopkg.in/check.v1"
 	"sync"
```
Comment (aanm) [inline daemon/daemon_test.go]: Imports order should be `golib`, `cilium`, `3rd parties`

`sync` should be on the first group

File: api/v1/openapi.yaml [hunk 5]
```diff
@@ -280,22 +280,22 @@ paths:
             "$ref": "#/definitions/Error"
   "/identity":
     get:
-      summary: Retrieve identity by labels
+      summary: Retrieves a list of identities that have metadata matching the provided parameters.
+      description: |
+        Retrieves a list of identities that have metadata matching the provided parameters, or all endpoints if no parameters provided.
```
Comment (ianvernon) [inline api/v1/openapi.yaml]: All endpoints? ;)

File: daemon/labels.go [hunk 6]
```diff
@@ -232,16 +233,33 @@ func NewGetIdentityHandler(d *Daemon) GetIdentityHandler {
...
 func (h *getIdentity) Handle(params GetIdentityParams) middleware.Responder {
-	lbls := labels.NewLabelsFromModel(params.Labels)
-	if id, err := LookupIdentityBySHA256(lbls.SHA256Sum()); err != nil {
-		if apierr, ok := err.(*apierror.APIError); ok {
-			return apierr
+	log.Debugf("GET /identity request: %+v", params)
+
+	identities := []*models.Identity{}
+	if params.Labels == nil {
+		outputList, err := kvstore.Client().ListPrefix(common.LabelIDKeyPath)
```
Comment (aanm) [inline daemon/labels.go]: Can you add a comment here? `// if labels is nil return all identities from the kvstore`

**Review by aanm — CHANGES_REQUESTED** [review]

File: tests/20-identity-list.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
```
Comment (ianvernon) [inline tests/20-identity-list.sh:1]: Use two spaces instead of four for indentation in the test scripts.

File: tests/20-identity-list.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+function restart_cilium {
+    echo "------ restarting cilium ------"
+    service cilium restart
+    echo "------ waiting for cilium agent get up and running ------"
+    wait_for_cilium_status
+}
+
+function cleanup {
+    gather_files 15-policy-config ${TEST_SUITE}
```
Comment (ianvernon) [inline tests/20-identity-list.sh]: Wrong test name for gather_files command.

File: tests/20-identity-list.sh [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+#
+# The test launches three containers and waits until 3 endpoints are created in Cilium. It then extracts the security ID
+# from the `cilium endpoint list` output.
+function test_identity_list {
+    remove_containers
+    restart_cilium
+    start_containers
+    wait_for_endpoints 3
+    cilium endpoint list
+    ID=$(cilium endpoint list | grep id.foo | awk '{print $3}')
```
Comment (ianvernon) [inline tests/20-identity-list.sh]: please make this a local variable

File: tests/20-identity-list.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    # Get expected response and replace all newline chars with a single space.
+    response=$(cilium identity list "container:id.foo" | sed ':a;N;$!ba;s/\n/ /g')
```
Comment (ianvernon) [inline tests/20-identity-list.sh]: please make this a local variable.

File: tests/20-identity-list.sh [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    echo "Endpoint security ID is: $ID"
+
+    expected_response='All non-reserved identities: (Note: If labels have been provided as parameters, only matching identities will be displayed) {   "Payload": [     {       "id": '$ID',       "labels": [         "container:id.foo"       ]     }   ] }'
```
Comment (ianvernon) [inline tests/20-identity-list.sh]: please make this local as well.

File: tests/20-identity-list.sh [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    if [[ "${expected_response}" != "${response}" ]]; then
+        abort "Expected: ${expected_response}; Got: ${response}"
+    fi
+}
+
+# Checks that the `cilium identity list --reserved` response matches expectations.
+function test_identity_list_reserved {
+    # Get expected response and replace all newline chars with a single space.
+    response=$(cilium identity list --reserved | sed ':a;N;$!ba;s/\n/ /g')
```
Comment (ianvernon) [inline tests/20-identity-list.sh]: please make this and line 66 local variables

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (ianvernon) [reply .gitignore:9]: Anything in `tests/cilium-files` is created by the tests and can be removed / ignored.

File: api/v1/openapi.yaml (hunk 5 above)
Comment (ashwinp) [reply api/v1/openapi.yaml]: Fixed.

File: tests/20-identity-list.sh (hunk 7 above)
Comment (ashwinp) [reply tests/20-identity-list.sh:1]: Fixed.

**Comment by ashwinp** [comment]
@ianvernon @aanm addressed all the review comments. I chose to ignore the test files in .gitignore as I think they might be useful for debugging test failures (therefore, we better not delete them as part of the cleanup).

Please review again.

**Review by ianvernon — APPROVED** [review]

File: cilium/cmd/identity_list.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func listIdentities(args []string) {
+	if reservedIDs {
+		fmt.Println("Reserved identities:")
+		for k, v := range policy.ReservedIdentities {
+			fmt.Printf("%3d %-15s \n", v, k)
+		}
+	}
+	fmt.Println("All non-reserved identities:\n" +
```
Comment (tgraf) [inline cilium/cmd/identity_list.go]: Should we label this:
> Identities in use by endpoints:

?

Most users may not understand what "non-reserved" means.

File: daemon/labels.go [hunk 14]
```diff
@@ -232,16 +233,34 @@ func NewGetIdentityHandler(d *Daemon) GetIdentityHandler {
...
+		lbls := labels.NewLabelsFromModel(params.Labels)
+		if id, err := LookupIdentityBySHA256(lbls.SHA256Sum()); err != nil {
+			if apierr, ok := err.(*apierror.APIError); ok {
+				return apierr
+			}
+			return apierror.Error(GetIdentityUnreachableCode, err)
+		} else if id != nil {
+			identities = append(identities, id.GetModel())
+		}
+		return NewGetIdentityOK().WithPayload(identities)
```
Comment (tgraf) [inline daemon/labels.go]: You could move the return out of the branch to avoid the return in both the `if` and `else` branch

**Review by tgraf — CHANGES_REQUESTED** [review]

File: api/v1/server/restapi/policy/get_identity_responses.go [hunk 15]
```diff
@@ -32,32 +32,35 @@ func NewGetIdentityOK() *GetIdentityOK {
...
 
 	rw.WriteHeader(200)
-	if o.Payload != nil {
-		payload := o.Payload
-		if err := producer.Produce(rw, payload); err != nil {
-			panic(err) // let the recovery middleware deal with this
-		}
+	payload := o.Payload
+	if payload == nil {
+		payload = make([]*models.Identity, 0, 50)
```
Comment (manalibhutiyani) [inline api/v1/server/restapi/policy/get_identity_responses.go:51]: Why do we hardcode 50 here? whats the rationale?

File: daemon/labels.go [hunk 16]
```diff
@@ -232,16 +233,34 @@ func NewGetIdentityHandler(d *Daemon) GetIdentityHandler {
 }
 
 func (h *getIdentity) Handle(params GetIdentityParams) middleware.Responder {
-	lbls := labels.NewLabelsFromModel(params.Labels)
-	if id, err := LookupIdentityBySHA256(lbls.SHA256Sum()); err != nil {
-		if apierr, ok := err.(*apierror.APIError); ok {
-			return apierr
+	log.Debugf("GET /identity request: %+v", params)
```
Comment (manalibhutiyani) [inline daemon/labels.go:236]: debug log left here by mistake?

**Review by aalemayhu — APPROVED** [review]
LGTM

File: api/v1/server/restapi/policy/get_identity_responses.go (hunk 15 above)
Comment (ianvernon) [reply api/v1/server/restapi/policy/get_identity_responses.go:51]: This code is automatically generated by Swagger - we do not control this

File: cilium/cmd/identity_list.go (hunk 13 above)
Comment (ashwinp) [reply cilium/cmd/identity_list.go]: Done

File: daemon/labels.go (hunk 14 above)
Comment (ashwinp) [reply daemon/labels.go]: Done

File: daemon/labels.go (hunk 16 above)
Comment (ashwinp) [reply daemon/labels.go:236]: not by mistake. I followed the style in endpoints.go. Moreover, this helped me while debugging.

**Review by tgraf — APPROVED** [review]

**Pull request merged**